### ISSUE TITLE
(v4) Possible to pull logger(s) into separate packages?

### ISSUE THREAD

**Issue by thekevinbrown** [mikro-orm/mikro-orm#658]
**Is your feature request related to a problem? Please describe.**
We're using Webpack to bundle Mikro. Currently 25% of our bundle size is `highlight.js`, which is being brought in by `cli-highlight`, which is used from Connection.ts, in `logQuery` in core. Chalk is also referenced here.

**Describe the solution you'd like**

Since in this environment we're not able to handle colours anyway (AWS Lambda), I was thinking it'd be nice if Mikro ORM loggers conformed to an ILogger interface like:

```typescript
interface ILogger {
  logQuery(query: string, took?: number, language?: string): void;
}
```

Then we could create two packages, `@mikro-orm/color-logger` and `@mikro-orm/plain-logger` which would only pull in the dependencies they need. Then a user could new up one of these loggers and pass it in instead of core needing to depend on chalk and cli-highlight itself.

**Describe alternatives you've considered**
Since we're not using this logging functionality, we could configure Webpack to use the null-loader to strip these out, but over time I think it'd be great to use dependency injection to allow users to customise Mikro anyway, outside of my use case. If the logger is configurable like this then users can use their own logging frameworks and not actually use either bundled logger package if they'd prefer.

**Additional context**
Happy to do this one if this sounds like a good idea.

**Comment by B4nan** [comment]
I was also looking into how large the highlight dependency actually is... I would say its fine to depend on chalk, but the highlighting could definitely be optional. 

I would prefer to have it in the repo so it can be enabled by default, but I agree its huge. Maybe we could implement custom highlighting function instead (we use it only for SQL basically, and I believe there is muuuuch more in that package, so we might just extract what is needed, or find some dependency free variant).

**Comment by thekevinbrown** [comment]
I'd really prefer to have all the color stuff including chalk not in the bundle, as we'll get genuinely zero benefit from having it there. This kind of goes back to that 'thin mikro' I want to work on. In general, the ideal for me would be

- Functionality that is distinct becomes separate packages, even down to loggers. There'd be a lot of packages.
- One of the packages is something like `@mikro-orm/default` which assembles all the pieces for the user in the automatic and recommended configuration. It would auto-discover entity schemas and all that jazz.
- If you want to assemble it differently, go ahead, just follow what's in the index file of the default package in your project but swap the pieces around as you need. If you know you're never going to use `@mikro-orm/ts-morph-discovery` then you don't even need to install it.

This gets the automatic behaviour you're describing for users, while also allowing users to better control or replace pieces through dependency injection that don't work in their use case.

Architecturally in general, the consistent thing I keep bumping into is Mikro wants to include everything and figure it all out on `init()`, while I want to tell it what I need it to do so it runs faster and is smaller. Right now init takes about a second on Lambda, which is a long time to wait on cold boot before we can even try connecting to the DB. It's guaranteed that we won't have changed the entities from build, so I'd like to be able to write the discovered entity info to a file at build time then just set it into Mikro.

Pulling the loggers out of core to me feels like a small step in that direction, and if it were up to me we'd take more steps in that direction so that we can get to the point where it's basically:

1. Discover Entities (using any of the ways that can work, ts-morph, whatever)
2. Instantiate loggers and any other pluggable logic you need
3. Pass those to init, which doesn't have any knowledge of how to discover anything anymore, it just goes with the objects you give it

**Comment by B4nan** [comment]
Without having `chalk` inside, how will you do partial coloring? We would need abstraction on top of chalk to do that, which sounds like wrong thing to do (not much benefit).

```typescript
this.logger.log('info', `MikroORM successfully connected to database ${chalk.green(db)}`);
```

**Comment by thekevinbrown** [comment]
Yeah, definitely don't want to add a layer on top of chalk. Agree that's the wrong thing to do. Chalk looks like it's the wrong example for what I'm really talking about.

**Comment by B4nan** [comment]
> @mikro-orm/default 

We already have the `mikro-orm` package (without the namespace), currently it is a meta package that requires others, I am not sure how to make it also automatically configure things (there are ways indeed).

> Architecturally in general, the consistent thing I keep bumping into is Mikro wants to include everything and figure it all out on init(), while I want to tell it what I need it to do so it runs faster and is smaller. Right now init takes about a second on Lambda, which is a long time to wait on cold boot before we can even try connecting to the DB. It's guaranteed that we won't have changed the entities from build, so I'd like to be able to write the discovered entity info to a file at build time then just set it into Mikro.

This sounds like you want metadata caching, and that is already available. That is the only way to get around discovery step. It's not enough to "compile things", it's not about TS vs JS, not about writing an "prepared entity" somewhere, it's about having all the metadata in the storage - types of properties and options of the relations. 

> There'd be a lot of packages.

I am not very bought on that idea of having 10 times (or whatever number) more packages and require users to wire everything (that sounds exactly like what I don't like about react, lol). The reason to separate something into a package (for me) is its dependencies being large, not (that much) the code itself. 

Btw it sounds quite weird that you have the discovery so slow, I am using metadata caching and it is much faster - just checking and I get 165ms in my production app (more than 20 quite large entities).

**Comment by thekevinbrown** [comment]
I can't cache it in memory because it's a lambda cold boot and our memory is blank each time. If the lambda container stays warm, it's definitely much faster, but there will be many users that need to get a response from a cold boot (I'm aiming for <3s total with ~1s being lambda time in this case). So caching isn't enough for us, we need the actual discovery in the case where there is no cache to be faster, or even better, not happen because I already supply entities that I discovered at build time. Is that something we can do that I've missed?

As far as wiring, that's what the default package would do. A normal user wouldn't need to mess with any of it or even know how many packages there are. Gatsby comes to mind in that they have [lots of packages](https://github.com/gatsbyjs/gatsby/tree/master/packages) but most people don't know or care that that's the case. The reason React is split up the way it is is precisely because people need to cut it up in all kinds of different ways. Their answer to this is create react app, which is very automatic, and we could have something like that for default Mikro users quite easily.

**Comment by B4nan** [comment]
Discovery is not a build step, but rather a runtime step, so nothing you can skip (as mentioned, you can skip it by using the caching). And while it is called discovery, it is not just about finding entities, it is about reading the metadata as well as filling in the defaults, and in the end also validating everything. 

Validation is something we could allow to be disabled to speed things up a bit, but I don't feel like it would save more then few milliseconds (I'd say less than 10ms). 

Btw you can provide your own caching adapter, so you could have a shared redis instance for this purpose, and that will definitely work with webpack, right?

**Comment by thekevinbrown** [comment]
If it's possible to cache what's needed in Redis, it should be possible to cache in serialised objects somewhere else, yeah? If so, couldn't I put Mikro through its paces on a build server, serialise the cache in the same way I'd serialise it to Redis, then save that to a file, then import that file at runtime, thus making it a build step?

Again this just feels like an idealogical difference. The Mikro approach in general seems to be "Mikro is a mostly singular big thing in control of all this stuff, you can provide an adapter to tweak this aspect of the behaviour, but it's still going to do all this other stuff automatically" where what I am trying to move closer to is "Mikro is a collection of packages that work together to achieve things and each package has a minimal responsibility. You can swap a package if you like by supplying a different package in the setup that was previously done automatically for you."

If that's absolutely not the direction you want to head, then that's fine, this issue becomes a simple one about how to swap syntax highlighting with another approach to make it smaller, and a different issue about how to write a caching adapter for my use case, was just hoping to move control away from a monolith into individual packages as that'd make it much easier to pull dependencies apart in general moving forward. V4 is a huge step forward in this regard (thank you!), just wanted to keep moving along in that direction if you're amenable.

**Comment by merceyz** [comment]
It might help to use https://yarnpkg.com/package/ansi-colors instead of chalk, it's smaller and faster

**Comment by thekevinbrown** [comment]
Yeah, on the caching, from reading https://mikro-orm.io/docs/metadata-cache/ I think it's as simple as:

- Create a CLI command that depends on ts-morph and Mikro.
- Run `init()` pointed at a DB with the same schema that we'll have on the final machine.
- Either deploy the ./temp folder with the webpacked output so it can be accessed at runtime, or (what I'm leaning towards) generate a specific cache adapter that includes the contents of the cache in ./temp within the actual webpack bundle.

I'll give that a go and will get back to you, but in general it'd be better for us at least if discovery was a separate process you could actively trigger, returning whatever Mikro needs from it as the result. Right now it's all lazy loaded across everything, where if you supplied three discovery packages and just let the user choose which one they wanted, it'd remove all of the need to automatically haul in things like ts-morph. The user could build their own, or could save the output of any of the built in packages, then could run this step anywhere they want.

Basically inverting the control and more moving distinct behaviour to separate packages that comply to a clear contract between them, similar to what I'm pushing for on the logging.

**Comment by thekevinbrown** [comment]
> Discovery is not a build step, but rather a runtime step...

I just found `mikro-orm cache:generate`. I think I finally understand what you were saying. I'd say "caching" isn't what's really happening here. Yes, you're caching the metadata, but you don't normally deploy a cache with the application. The word "caching" and mention of Redis lead me to believe that I'd have to do this as part of the application and that it was impossible to do at build time, while actually you already have a CLI command for doing exactly that.

**Comment by B4nan** [comment]
> I'll give that a go and will get back to you, but in general it'd be better for us at least if discovery was a separate process you could actively trigger, returning whatever Mikro needs from it as the result. Right now it's all lazy loaded across everything, where if you supplied three discovery packages and just let the user choose which one they wanted, it'd remove all of the need to automatically haul in things like ts-morph. The user could build their own, or could save the output of any of the built in packages, then could run this step anywhere they want.

Not sure why you say so, discovery is distinct process and it is not lazy (not in the sense that things are executed when needed) and spread across everything, the whole discovery always happens in the init phase, and you can trigger it manually:

```typescript
const discovery = new MetadataDiscovery(MetadataStorage.init(), orm.driver.getPlatform(), orm.config);
orm.metadata = await discovery.discover();
```

**Comment by thekevinbrown** [comment]
Ok, that was a reaction to

> Discovery is not a build step, but rather a runtime step, so nothing you can skip

Which I took to mean it has to happen on the main server and can't be triggered explicitly. It's helpful to understand how Mikro is assembled. I'm still struggling to see all the layers and how they interact, so if I'm asking dumb questions it's because I don't understand what you've built. We were using the ReflectMetadata provider before so I had no idea the cache was even a thing, or that you'd want to deploy it with the entities.

**Comment by B4nan** [comment]
Maybe we could replace highlight.js with this (also zero dependencies):

https://github.com/scriptcoded/sql-highlight

In fact that package is basically single file with less than 100loc, so we might even inline it to have absolute control.

Or inline it, keeping only the SQL + JS support (JS support is used for mongo queries). Also CLI uses TS language when highlighting generated migration file. 

Overall I would like to have this implicitly enabled, not as a separate package.

Btw I already swapped `chalk` with `ansi-colors`.

**Comment by thekevinbrown** [comment]
I'd vote for sql-highlight.

**Comment by B4nan** [comment]
Already did a try with that, I might have screwed something on the way (I did inline it and cleaned up the code quite a lot), but it was not working very well (most of the queries, even the simple ones, were not just incorrectly highlighted, but also messed up/modified, so the actual string was different, not just colors). 

Will give it another try later this week I hope.

**Comment by B4nan** [comment]
Following up on #676:

Regarding the coloring, it is just about the highlighting, right? Not about the chalk usage I hope (now replaced with ansi-colors that should be much faster). It can be already disabled via configuration (from the very beginning of this addition), does that help? You can also disable coloring via `process.env.FORCE_COLOR = 0`.

You mentioned that the issue is in the `init` phase, where we basically do not color anything (the whole discovery gives us just few logs without highlighting being even triggered... So this really feels like the whole problem is the highlight.js and its usage/registration. Can you verify that?

**Comment by thekevinbrown** [comment]
I’ll verify and will get you a PR tomorrow that fixes it, so no need to worry, leave it with me.

**Comment by thekevinbrown** [comment]
Ok, so the 500ms on init was caused by simply requiring `cli-highlight`, since webpack does evals when a require happens. 

The other thing I discovered is that about 75% of the time spent running queries in the default configuration was spent colorising log output that was not actually logged. By default highlight is true (so it does the highlighting logic when a query happens) but debug is false (so there's no actual output). #680 fixes this so that if there's not going to actually be a log it doesn't bother to colorise it.

For the highlight side of it, right now the approach is to pull in everything, then switch behaviour on and off with flags. I still feel like logging is a good use case for separating into packages, but that's more of an architectural change. We can supply a shim for cli-highlight that looks like this:

```javascript
module.exports = {
	fromJson: () => {},
	Theme: {},
};
```

Then do this in webpack config:

```
resolve: {
		extensions: ['.ts', '.js'],
	alias: {
		// We don't use cli-highlight at runtime and it's huge, so
		// there's no need to include it.
		'cli-highlight': path.resolve('src', 'cli-highlight-stub'),
	},
},
```

But you can see how that'll be error prone in the long term, as it's quite tightly coupled to the internals of Mikro. I think rather than doing surgery on Mikro and cutting it apart with webpack it'd be better to switch to more of a pattern of only bringing in what's needed. I'll work on a PR for this after the fulltext operator is complete, as this works for now.

**Comment by B4nan** [comment]
Yeah we need to do something about this in general, not everyone is using webpack to be able to get around it like this.

I think my problem with the logger as separate package proposal is that in fact the issue is not with logging, but with highlighting. Maybe we should have `@mikro-orm/highlighting` as a package, but still I would prefer to get rid of highlight.js and implement something lighter, having it enabled by default. Will try the `sql-highlight` one more time as that could really be inlined and kept all the time, also want to take a look at https://github.com/PrismJS/prism.

**Comment by B4nan** [comment]
PrismJS is also fat, so that would not help much. the `sql-highlight` is purely regexp based, which is in general not a good approach, but rather just an approximation... 

Maybe we could develop custom highlighting based on some existing sql parser, like this one: https://github.com/forward/sql-parser

(this one has also zero dependencies)

**Comment by thekevinbrown** [comment]
I've thought about this quite a bit, and it is actually about logging, not just highlighting. When color control codes come out in the logs it breaks our logging output in CloudWatch. So if we want to give users control here and approach it like you're saying, we really need 3 layers:

- Highlighter
- ColorProvider
- LoggingProvider (already exists)

And at that point core knows quite a bit about how to build up logging messages. What if a user wants to log certain messages to CloudWatch and others to a file? As a user I then need to understand the surface area of 3 providers, various core settings (highlight: false overrides what your highlighter does) as well as what core is doing to glue it all together just to be able to change how the logging works. That's really hard to customise.

Wouldn't it be better to have one API that core uses to log things, then have 3 supported states: a LoggingProvider of null which means don't log, a SimpleLoggingProvider which just puts uncolorised strings out on the console, and a HighlightingLoggingProvider which colors them up? The API here wouldn't need to be very complex, I've looked at how the colorising is working and it'd be pretty simple to pass enough info down to the provider so it can make coloring decisions. It'd then be much easier for users to integrate with their own logging framework like Bunyan, as they could see that they just replace this one provider and they're good to go.

Otherwise we're going back to "the core brings in dependencies you can't use in your environment, spends cycles applying colors, then you can strip them back out after that's happened", which is what I am trying to get us away from. I get it's not a huge performance impact, I just think it'd be better to allow the user to decide how they want the logging to work instead of having to completely understand 3 separate layers and several settings just to customise the logging behaviour.

**Comment by B4nan** [comment]
I thought we already agreed that abstracting chalk/ansi-colors is not important/worth it. You can disable it globally, and I really doubt it has a real performance impact (unlike huge highlight.js & friends). 

My main concern here is that this would literally make the installation and configuration more complicated (and so far I am not bought on the idea of `@mikro-orm/default` - I guess mainly because I really don't like the name 🤷). I can see the need for separating highlighting (and again, I would really like to see a working, dep-free default highlighter in core so we do not force users to configure this - similar to how metadata providers work), but just for coloring? I really do not see the benefit here, I doubt the coloring we do (not highlighting of queries) has a real performance impact (especially after moving to `ansi-colors`).

> As a user I then need to understand the surface area of 3 providers, various core settings (highlight: false overrides what your highlighter does, debug: false means don't log at all, etc) as well as what core is doing to glue it all together just to be able to change how the logging works. That's really hard to customise.

Idk, for me the existing API is far from "hard to understand". You have a way to disable highlighting and a way to disable debug mode/logging (and you need to explicitly enable it yourself first). You can provide a logger of yours, so that is fully customizable. You can disable coloring too (globally, we could have a toggle for that if needed). You can even change the highlighting theme. For me this really allows everything we need, and my only concern is with performance.

> Wouldn't it be better to have one API that core uses to log things, then have 3 supported states: a LoggingProvider of null which means don't log, a SimpleLoggingProvider which just puts uncolorised strings out on the console, and a HighlightingLoggingProvider which colors them up? The API here wouldn't need to be very complex, I've looked at how the colorising is working and it'd be pretty simple to pass enough info down to the provider so it can make coloring decisions. It'd then be much easier for users to integrate with their own logging framework like Bunyan, as they could see that they just replace this one provider and they're good to go.

I do not see how you could have `HighlightingLoggingProvider` - loggers do not know about highligting, they already receive (possibly highlighted) strings. 

Looking at some other projects, both knex and nestjs do have a direct dependency on something like chalk.

**Comment by B4nan** [comment]
> but what about users who want to alter the behaviour beyond just turning things on and off?

We were talking about coloring and highlighting, and as I mentioned, you can also customize the theme too. I fail to imagine what more control would anyone want there. 

> There are many things users may want to do that are more than on or off, for example logging queries to a file, but errors to CloudWatch. How would they do that currently? We can easily give them these options by structuring the packages so it's easier to untangle and they can then more easily plug in their own behaviours.

How? You pass custom instance of logger method, and do whatever you want from there, based on the log namespace. How would this be better if we have separated logger into its own package? I don't see a difference. 

> A good example for this type of thing is database drivers. SQLite, Mongo, and Postgres all take quite different connection parameters. Right now, the core Mikro config has to understand all of these variations. What if I want to add a DynamoDB driver? Why should I have to change core to add it? As long as it complies with the same interface the others do, core shouldn't care that there's a new one, but right now it does.

Not really, you have absolute control over what is passed to the driver via `driverOptions`. Also I don't understand where you see the core needs to understand the params and do something about them. You can create any driver, including dynamodb, without touching the core (nothing changed, this was true in v2 and v3 too). The abstraction is there, it's the `IDatabaseDriver` you need to conform, so I really don't understand what you mean by this. Or which part of core you think you would need to adjust?

> Would me building a HighlightingLoggingProvider help? It's a change to the way logging providers work, but honestly you already have the non-highlighted API already, it's just in Connection called logQuery for example. The logging providers would take on more responsibility, while core knows nothing about logging other than to call logQuery, log[whatever], thus making it easier to untangle, and making the responsibility very clear, LoggingProviders handle logging. Core calls these specific methods on the logging providers and doesn't know and/or care how that works at the next layer down.

I am still not sure what you mean by that - what would the interface look like? Highlighting also takes place in the cli package, its not just about logging queries (and not just about SQL).

**Comment by B4nan** [comment]
So I ended up rewriting https://github.com/doctrine/sql-formatter to typescript (tokenizer and the highlighting part), as apparently there is no good enough solution (designed for node and not browser) in the whole JS ecosystem :D

It's working quite good, need to do some adjustments before I ship it. Will keep it independency of the rest of MikroORM packages so others can use it.

**Comment by thekevinbrown** [comment]
That's great.

I still feel like pulling logging out and treating it as a separate package from the core would be an improvement, but have not had time to answer your questions. Will see if I can get to it this week.

**Comment by B4nan** [comment]
Dependency on `cli-highlight` (and therefore `highlight.js` as well) is now removed via #718. I will close this one, as I really don't feel like it is worth it to make the `ansi-colors` optional. Its a dependency free package, and its really tiny (and it can be disabled globally)...

But will be happy to hear concrete arguments on why it would be an improvement. In my eyes it would not even save CPU cycles/event loop ticks - because we would need to wrap the usage, so the code would probably end up being more complex than using the library directly.
